### ISSUE TITLE
YASL strings are output with respect to null-terminators, should be with respect to length

### ISSUE THREAD

**Issue by CoffeeTableEspresso** [yasl-lang/yasl#355]
YASL does not use C-strings, this is a leaky implemetnation detail. Pritning strings should show the whole string, not just up to the first nul character.

**Comment by DarthUdp** [comment]
May I have a stab at this one?

**Comment by CoffeeTableEspresso** [comment]
@DarthUdp Certainly!

I can point you in the right direction too.

The offending line of code is found in `interpreter/VM.c`, in the `vm_ECHO` function.

This calls into `vm_print_out`, which uses the same kind of formatting as `printf`. This doesn't handle the NUL characters properly.

The fix I have in mind is to add another function for `struct IO` (see: `IO.h`), that takes a string and a length (instead of doing formatting).

There's no need to worry about nicely formatting the output (like escaping `\0` or whatever), just print out the right number of bytes.

If you run into any trouble, let me know!

**Comment by DarthUdp** [comment]
alright, I will have a go at it over the weekend, thanks for the pointers I will let you know if I hit any walls, don't expect to but you can never be sure

**Comment by DarthUdp** [comment]
Some findings I made, the function called by the print statement of the vm is actually io_print_file and not io_print_string, had to use a debugger to find out wth any changes I made to io_print_string didn't have any effects.

**Comment by DarthUdp** [comment]
Also, all formatting functions that take a length are length or first null in their behavior

**Comment by DarthUdp** [comment]
very dirty and hack as hell implementation, that keeps the signature and will print whatever you give it as a nice bonus shows where nulls are, we might want to discuss new signatures for this part of the code.
Do you need the format string? I'd say since it's a fixed value anyways it's doing more harm than good there.

```c
/**
 * To avoid mucking about with the build process
 */
void nop(void *val) {
	asm("nop":: "r"(val));
}

/**
 * to summarize: this is not the function intended to print strings I guess?
 * but every formatting function in C that takes a length will stop on the first null thus we need to revise this
 * whole call stack to stop relying on the format
 */
void io_print_file(struct IO *const io, const char *const format, va_list args) {
	nop((void *)format);
	nop(io);
	va_list args_copy;
	va_copy(args_copy, args);
	size_t len = va_arg(args_copy, size_t);
	char *str = va_arg(args_copy, char *);
	va_end(args_copy);
	for (size_t i = 0; i < len; i++) {
		if ((int)str[i] == 0) {
			printf("\\0");
		}
		putchar((int)str[i]);
	}
	putchar('\n');
}
```

**Comment by CoffeeTableEspresso** [comment]
Yes, it's possible to configure whether outputs are printed to a file (`stdout` for example), or a string. (Hence, `io_print_file` and `io_print_string`.) By default it prints to a file but either way is possible depending on what's set up.

These two are also used for printing error messages, so changing them (rather than just adding a length-based alternative) may leave you a bit of work. The error messages should be fine for the most part, but as you point out I probably have a few more issues with handling NULs inside strings elsewhere in YASL.

**Comment by DarthUdp** [comment]
so you do use the format for printing errors in a nicer way?

**Comment by CoffeeTableEspresso** [comment]
Yes, the format is used widely for printing errors. `vm_print_err` is the specific function that calls into this for runtime errors, as an example.

**Comment by DarthUdp** [comment]
maybe make echo into a function or function-like then or create something like print so this part can be left alone, because this is quite a deep hole that maybe is best left alone or we write a `printf` family clone that is null tolerant and length oriented, doable but I would simplify the formatting to what we strictly need for errors (I'm thinking one that outputs to a buffer `s` and escapes/removes nulls so we can then use the old way to output it)

**Comment by DarthUdp** [comment]
or even at first a function to just strip all nulls before even sending it to IO code

**Comment by CoffeeTableEspresso** [comment]
Yes, what I'd like to do is have a separate function for IO used internally that just takes a length and a string and "prints" it. And just completely leave all the formatting stuff needed for errors alone.

**Comment by DarthUdp** [comment]
Got a solution will open a pr in a minute

**Comment by CoffeeTableEspresso** [comment]
fixed in v0.11.2
